src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +304,54 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    /// <summary>
+    /// Creates a <see cref="LoadedProject"/> which has bare minimum information, but, which documents can be added to and obtained from.
+    /// </summary>
+    protected LoadedProjectSet AddLoadedProjectSet(string projectPath)
+    {
+        var projectSet = new LoadedProjectSet([], new SemaphoreSlim(1), new CancellationTokenSource());
+        return _loadedProjects.GetOrAdd(projectPath, projectSet);
```

Comment:
It might be reasonable to use TryAdd and assert that we were actually the ones who added this.
- Only LS queue should be adding new entries here.
- We can see that when `TryUnloadProjectSetAsync` completes, the key `projectPath` *must* be gone from `_loadedProjects`.
- Effectively we do not expect different threads calling Add to race with each other, or for threads calling Remove and Add to be racing with each other.
- It's not clear what we should do here, if the project set already exists--someone might be adding a project to it, kicking off design time build etc.